### PR TITLE
feat(#319): add optional change_type_order

### LINKED ISSUE #319: [feature/changelog] Order Change Types Based on Bump Map Rules
## Description

The `CHANGELOG` is ordered with the most recent change type first. This makes the diffs hard to read and may put less important change types over more important ones

Example diff from one of my projects: https://github.com/KyleKing/calcipy/commit/a61d0a4447f5ec03a3eccf279533be7955801256#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed

## Possible Solution

Could the `CHANGELOG` be ordered based on the bump rules?

The `CHANGELOG` could be ordered in descending importance from `(MAJOR, MINOR, PATCH, OTHER)` where the types within each section would be sorted alphabetically and the `OTHER` section would capture any change types shown in the `CHANGELOG`, but not part of the bump rules

## Related Issue

I looked through the issues (matching [order](https://github.com/commitizen-tools/commitizen/issues?q=order) or [changelog](https://github.com/commitizen-tools/commitizen/issues?q=changelog)) but didn't find any that appear to be asking for the same feature

## PR Notes

I would be happy to try to add this functionality, if this feature would be of interest to the maintainers. If not, I may look into separate [customization](https://commitizen-tools.github.io/commitizen/customization/#custom-changelog-generator) as an alternative to standard functionality) 

I created a [cz_legacy customization](https://github.com/KyleKing/cz_legacy) last night, so I have some familiarity with the commitizen code base

### PR DESCRIPTION
**Pull request by KyleKing**
## Description

Implements proposed feature in #319. The default change type order is `["BREAKING CHANGE", "feat", "fix", "refactor", "perf"]` (in descending order of importance) followed by any other change types in alphabetical order. In the configuration file, a user can override this setting to specify their own order and cover custom change types of interest.

## Checklist

- [X] Add test cases to all the changes you introduce
- [X] Run `./script/format` and `./script/test` locally to ensure this change passes linter check and test
- [X] Test the changes on the local machine manually
- [X] Update the documentation for the changes

## Expected behavior

The `CHANGELOG` should be ordered in the user-specified order rather than FIFO

## Steps to Test This Pull Request

Run the standard pytest suite and see the two new test cases

## Additional context

Fixes #319

Example use: https://github.com/KyleKing/calcipy/commit/4a480a8ca91350c1a4c190a672af1e9d397a6eb2

### CODE REVIEW

**Comment by KyleKing** [comment]
I think this is all done. I cleaned up the commit history so that the first commit is just a refactor to move the commit tree to a global variable for use by my new tests. The diff on the second commit are the actual changes: https://github.com/commitizen-tools/commitizen/pull/323/commits/e855269045207aa09e3c9be271aec05be9ef39e7

I've tested the changes in my personal project and all seems to work well even with custom change types: https://github.com/KyleKing/calcipy/commit/4a480a8ca91350c1a4c190a672af1e9d397a6eb2

**Comment by Woile** [comment]
Hi KyleKing, awesome work here! Minor correction, this is a new feature, so I'd like to ask you to rebase your PR using `feat` instead of `fix`, or if you want I can squash it and rename it, no problem.

**Comment by KyleKing** [comment]
@Woile, thanks! Good point 🤦 , I amended the commit message

File: docs/customization.md [hunk 1]
```diff
@@ -51,7 +52,7 @@ The equivalent example for a json config file:
 ```json
 {
     "commitizen": {
-	"name": "cz_customize",
+    "name": "cz_customize",
```
Comment (Lee-W) [inline docs/customization.md:55]: I think we might still need this indentation.

File: tests/test_changelog.py [hunk 2]
```diff
@@ -509,293 +509,319 @@ def test_get_commit_tag_is_None(gitcommits, tags):
...
+                3: (["BREAKING CHANGE", "refactor"], ["refactor", "BREAKING CHANGE"]),
+            },
+        ),
+    ),
+)
+def test_order_changelog_tree(change_type_order, expected_reordering):
+    tree = tuple(changelog.order_changelog_tree(COMMITS_TREE, change_type_order))
+
+    index_of_reordered_entry = [*expected_reordering.keys()]
+    for index, entry in enumerate(tuple(tree)):
```
Comment (Lee-W) [inline tests/test_changelog.py:816]: As `tree` is already a `tuple` on line 813, do we need the `tuple()` here?

File: tests/test_changelog.py [hunk 3]
```diff
@@ -509,293 +509,319 @@ def test_get_commit_tag_is_None(gitcommits, tags):
...
+
+
+@pytest.mark.parametrize(
+    "change_type_order, expected_reordering",
+    (
+        ([], {}),
+        (
+            ["BREAKING CHANGE", "refactor"],
+            {
+                2: (["refactor", "feat", "fix"], ["feat", "fix", "refactor"]),
```
Comment (Lee-W) [inline tests/test_changelog.py:806]: I think we might need some comments on the magic number `2` and `3`, which represent `v1.1.0` and `v1.0.0`

File: tests/test_changelog.py [hunk 4]
```diff
@@ -509,293 +509,319 @@ def test_get_commit_tag_is_None(gitcommits, tags):
...
+                3: (["BREAKING CHANGE", "refactor"], ["refactor", "BREAKING CHANGE"]),
```
Comment (Lee-W) [inline tests/test_changelog.py:807]: How about making `expected_reordering` a dict of dict? (e.g., `{2: {"sorted_order": ..., "original_order": ...}}`) It took me sometime to understand this test

File: commitizen/changelog.py [hunk 5]
```diff
@@ -128,6 +129,19 @@ def generate_tree_from_commits(
...
 
 
+def order_changelog_tree(tree: Iterable, change_type_order: List[str]) -> Iterable:
+    sorted_tree = []
+    for entry in tree:
+        entry_change_types = sorted(entry["changes"].keys())
+        ordered_change_types = []
+        for ct in change_type_order + entry_change_types:
+            if ct in entry_change_types and ct not in ordered_change_types:
+                ordered_change_types.append(ct)
```
[the remark this answers is not in the document]
Comment (KyleKing) [reply commitizen/changelog.py:139]: Thanks for the feedback. I don't think sets preserve order, so I'll need to check if there are any edge cases and try out the refactor. I would like to make this code more concise

I'll probably try to push a commit with changes by Saturday

[the remark this answers is not in the document]
Comment (KyleKing) [reply commitizen/changelog.py:139]: Oh I see, I misread how your snippet would be used

I created a script to test the refactoring options. In the case that the user specifies duplicate keys, I think commitizen should raise an exception (`param` #3 below). If duplicates aren't cleaned from user input, sections would be repeated in the CHANGELOG. Additionally, all keys not in the commit history will need to be filtered after this sorting. I'll implement your suggestion and raise an exception on configuration errors

```py
"""Compare refactoring options."""

from typing import Callable, Iterable, List


def option_0(entry_keys: Iterable, change_type_order: List[str]) -> Iterable:
    """Variation of original implementation."""
    entry_change_types = sorted(entry_keys)
    ordered_change_types = []
    for ct in change_type_order + entry_change_types:
        # if ct in entry_change_types and ct not in ordered_change_types:
        if ct not in ordered_change_types:
            ordered_change_types.append(ct)
    return ordered_change_types


def option_1(entry_keys: Iterable, change_type_order: List[str]) -> Iterable:
    """Proposed alternative."""
    ordered_change_types = change_type_order + sorted(
        set(entry_keys) - set(change_type_order)
    )
    return ordered_change_types


def option_2(entry_keys: Iterable, change_type_order: List[str]) -> Iterable:
    change_types = change_type_order + sorted(entry_keys)
    return [ct for idx, ct in enumerate(change_types) if ct not in change_types[:idx]]


def test_option(func: Callable) -> None:
    param = [
        # Standard Use Case
        (['fix', 'feat', 'new'], ['new', 'fix'], ['new', 'fix', 'feat']),
        # No Overlap
        (['perf', 'fix', 'feat'], ['new', 'old'], ['new', 'old', 'feat', 'fix', 'perf']),
        # Duplicate in User-Specified
        (['perf', 'fix', 'feat'], ['new', 'alt', 'new'], ['new', 'alt', 'feat', 'fix', 'perf']),
    ]
    for idx, (entry_keys, change_type_order, expected) in enumerate(param):
        result = func(entry_keys, change_type_order)
        assert result == expected, '{} / {}'.format(param[idx], result)


if __name__ == '__main__':
    for func_name in [key for key in globals().keys() if key.startswith('option_')]:
        try:
            test_option(globals()[func_name])
        except AssertionError as error:
            print('Error on {}: {}'.format(func_name, error))
        else:
            print('Completed {}'.format(func_name))
```

**Review by KyleKing — COMMENTED** [review]
I restored the white space changes and I think I addressed all of the comments. Thanks for the feedback!

File: commitizen/changelog.py [hunk 6]
```diff
@@ -130,14 +130,21 @@ def generate_tree_from_commits(
 
 
 def order_changelog_tree(tree: Iterable, change_type_order: List[str]) -> Iterable:
+    if len(set(change_type_order)) != len(change_type_order):
+        raise RuntimeError(
```
Comment (Lee-W) [inline commitizen/changelog.py:134]: We prefer to make all the known exceptions as `CommitizenException` (https://github.com/commitizen-tools/commitizen/blob/master/commitizen/exceptions.py#L28). Could you please create one, add an exit code and update the documentation here? https://github.com/commitizen-tools/commitizen/blob/master/docs/exit_codes.md

**Review by Lee-W — COMMENTED** [review]
@KyleKing Thanks for your rapid update! There's only one minor exception improvement I'd like to have. I think we're almost ready to merge this one! 🎉

File: commitizen/changelog.py (hunk 6 above)
Comment (KyleKing) [reply commitizen/changelog.py:134]: Okay! I added a general `InvalidConfigurationError` for configuration errors like this with a new exit code (I will also add a custom exception to `cz_legacy` as well)

**Review by Lee-W — APPROVED** [review]
@KyleKing Thanks for the latest update! Let's merge it back to master

**Pull request merged**